R/property.R

```diff
@@ -84,66 +87,110 @@ new_property <- function(name, class = NULL, getter = NULL, setter = NULL) {
 #' ))
 #' lexington <- horse(colour = "bay", height = 15)
 #' lexington@colour
-#' property(lexington, "colour")
+#' prop(lexington, "colour")
 #'
 #' lexington@height <- 14
-#' property(lexington, "height") <- 15
+#' prop(lexington, "height") <- 15
 #'
-#' try(property(lexington, "age"))
-#' property_safely(lexington, "age")
- property <- function(object, name) {
-  val <- property_safely(object, name)
-  if (is.null(val)) {
+#' try(prop(lexington, "age"))
+#' prop_safely(lexington, "age")
+prop <- function(object, name) {
+  if (!inherits(object, "R7_object")) {
+    stop("`object` is not an <R7_object>")
+  } else if (!prop_exists(object, name)) {
     class <- object_class(object)
-    stop(sprintf("Can't find property %s@%s", fmt_classes(class@name), name), call. = FALSE)
+    stop(sprintf("Can't find property %s@%s", fmt_classes(attr(class, "name")), name))
+  } else {
+    prop_val(object, name)
   }
-
-  val
 }
 
-#' @rdname property
+#' @rdname prop
 #' @export
-property_safely <- function(object, name) {
+prop_safely <- function(object, name) {
   if (!inherits(object, "R7_object")) {
-    return(NULL)
+    NULL
+  } else if (!prop_exists(object, name)) {
+    NULL
+  } else {
+    prop_val(object, name)
   }
+}
+
+# Internal helper that assumes the property exists
+prop_val <- function(object, name) {
   if (identical(name, ".data")) {
     # Remove properties, return the rest
-    props <- properties(object)
-    for (name in names(props)) {
+    for (name in prop_names(object)) {
       attr(object, name) <- NULL
     }
+
     obj_cls <- object_class(object)
     class(object) <- setdiff(class_names(obj_cls@parent), obj_cls@name)
     object_class(object) <- object_class(obj_cls@parent)
+
     return(object)
   }
+
   val <- attr(object, name, exact = TRUE)
   if (is.null(val)) {
-    prop <- properties(object)[[name]]
+    prop <- prop_obj(object, name)
     if (!is.null(prop$getter)) {
       val <- prop$getter(object)
     }
   }
   val
 }
 
-properties <- function(object) {
-  obj_class <- object_class(object)
-  prop <- list()
-  while(!is.null(obj_class)) {
-    prop <- c(attr(obj_class, "properties"), prop)
-    obj_class <- attr(obj_class, "parent", exact = TRUE)
+# Get underlying property object from class
+prop_obj <- function(object, name) {
+  class <- object_class(object)
+  attr(class, "properties")[[name]]
+}
+
+#' @rdname prop
+#' @export
+prop_names <- function(object) {
+  if (inherits(object, "R7_class")) {
+    names(attributes(object))
```

Comment:
I'm not 100% sure this is correct; I just intuited my way to a solution. I think it's something to do with the way we have bootstrap a R7 class out of nothing, or possibly because there's two meanings of properties: the "formal" properties (i.e. the spec) and the actual properties.